### PR TITLE
Add support for Unix shebangs

### PR DESCRIPTION
**Pull request by Gabriel439**
… as standardized in https://github.com/dhall-lang/dhall-lang/pull/1158

### CODE REVIEW

File: dhall/src/Dhall/Parser/Expression.hs [hunk 1]
```diff
@@ -757,7 +766,7 @@ parsers embedded = Parsers {..}
                   where
                     predicate c = c == '$' || c == '\''
 
-                endOfLine = do
+                endOfLine_ = do
                     a <- "\n" <|> "\r\n"
```
Comment (TristanCacqueray) [inline dhall/src/Dhall/Parser/Expression.hs:770]: nit: could this be replaced by `a <- endOfLine` ?

**Comment by basile-henry** [comment]
It would be nice to support multiple lines of `shebang` so that we could use the `nix-shell` shebangs.

I have just tested this branch with the following (from the root of the `dhall-haskell` repo):

```dhall
#! /usr/bin/env nix-shell
#! nix-shell -i "dhall --file" -p "(import ./.).dhall"

42
```

Which currently gives me:

```bash
➜ ./test.dhall
dhall:
Error: Invalid input

./test.dhall:2:1:
  |
2 | #! nix-shell -i "dhall --file" -p "(import ./.).dhall"
  | ^^^^^^^^
unexpected "#! nix-s"
expecting expression or whitespace
```

I saw the `dhall-lang` PR for this, but didn't think about mentioning the `nix-shell` shebang use case at the time :sweat_smile:

**Comment by Gabriel439** [comment]
@basile-henry: Yeah, that would first require an upstream change to the standard, which we can still do afterwards

**Comment by TristanCacqueray** [comment]
It seems like `nix-shell` shebang is relying on the fact that `#` is a comment in the target language, but that does not always work. Though that seems like a good thing to support.

Alternatively, perhaps nix-shell could be modified to also consider `--! nix-shell -i "dhall --file" -p dhall` as a valid extra shebang line, but that would not look good.

**Comment by TristanCacqueray** [comment]
Also it is unfortunate that `#!/usr/bin/env -S nix-shell -p dhall -i "dhall --file"` does not work.

**Comment by basile-henry** [comment]
> `nix-shell` shebang is relying on the fact that `#` is a comment in the target language

That's probably true for `bash` and `python`. However I have seen multi-line shebangs with `nix-shell` for Haskell too, usually with GHC's `runhaskell`. That probably means that GHC takes this edge case specifically into account?

**Review by basile-henry — APPROVED** [review]

**Comment by TristanCacqueray** [comment]
`runhaskell` does something special where it ignores any lines starting with '#!', not just the first ones, according to: https://gitlab.haskell.org/ghc/ghc/-/blob/e61d2d47c4942c829ec98016be5aa5ae36982524/compiler/GHC/Parser/Lexer.x#L279

**Pull request merged**